File: algorithms/456.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

class Solution {
public:
    bool find132pattern(vector<int> &nums) {
        int n = nums.size();
        int s3 = INT_MIN;
        stack<int> s;
        for (int i = n-1; i >= 0; i--) {
            if (nums[i] < s3) return true;
            while (!s.empty() && s.top() < nums[i]) {
                s3 = s.top();
                s.pop();
            }
            s.push(nums[i]);
        }
        return false;
    }
};

int main () {
    vector<int> nums = {1, 5, 0, 3};
    Solution solution;
    solution.find132pattern(nums);
    return 0;
}

File: algorithms/rust/456.rs
impl Solution {
    #[allow(dead_code)]
    pub fn find132pattern(nums: Vec<i32>) -> bool {
        let mut stack = vec![];
        let mut third = i32::MIN;
        for &num in nums.iter().rev() {
            if num < third {
                return true;
            }
            while let Some(&k) = stack.last() {
                if k >= num {
                    break;
                }
                third = k;
                stack.pop();
            }
            stack.push(num);
        }
        false
    }
}

struct Solution;

fn main() {
    println!("Hello, world!");
}
